KAFKA-3770: KStream job should be able to specify linger.ms

gfodor opened the pull request:
See https://issues.apache.org/jira/browse/KAFKA-3770

ewencp committed:
KAFKA-3605: Return error if connector config includes mismatching connector name.

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Jason Gustafson

Closes #1253 from ewencp/kafka-3605-connector-name-mismatch

dguy committed:
MINOR: add unit test for KGroupedTable.count

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Michael G. Noll <michael@confluent.io>

Closes #1255 from dguy/kgroupedtable-count-test

gfodor committed:
KAFKA-3619: File handles are leaked on .lock files of ProcessorStateManager

Kafka Streams seems to hold file handles on the `.lock` files for the state dirs, resulting in an explosion of filehandles over time. Running `lsof` shows the number of open filehandles on the `.lock` file increasing rapidly over time. In a separate test project, I reproduced the issue and determined that in order for the filehandle to be relinquished the `FileChannel` instance must be properly closed. Applying this patch seems to resolve the issue in my job.

Author: Greg Fodor <gfodor@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1267 from gfodor/bug/state-lock-filehandle-leak

benstopford committed:
KAFKA-3270; Added some Happy Path Tests for the Reassign Partitions Command

with help from enothereska  :)

Author: Ben Stopford <benstopford@gmail.com>

Reviewers: Jun Rao <junrao@apache.org>, Eno Thereska <eno.thereska@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #956 from benstopford/KAFKA-3270-ReassignPartitionsCommand-Tests

hachikuji committed:
KAFKA-3602; Rename RecordAccumulator dequeFor() and fix usage

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Grant Henke <granthenke@gmail.com>, Ashish Singh <asingh@cloudera.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1254 from hachikuji/KAFKA-3602

guozhangwang committed:
KAFKA-3499: prevent array typed keys in KeyValueStore

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Ismael Juma, Josh Gruenberg, Michael G. Noll, Ewen Cheslack-Postava

Closes #1229 from guozhangwang/K3499

jklukas committed:
MINOR: Fix some copy-pasted Javadoc in StreamsConfig.java

This contribution is my original work and I license the work to the Kafka project under the project's open source license.

cc guozhangwang miguno ymatsuda

Author: Jeff Klukas <jeff@klukas.net>

Reviewers: Michael G. Noll <michael@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

Closes #1270 from jklukas/streams-doc-fix

Commit:
KAFKA-3614: Consolidate duplicate code in KGroupedTableImpl

Feel free to review guozhangwang enothereska mjsax .

Author: Michael G. Noll <michael@confluent.io>

Reviewers: Matthias J. Sax, Michael G. Noll, Eno Thereska

Closes #1262 from miguno/KAFKA-3614

mjsax committed:
KAKFA-3599: Move WindowStoreUtils to package "internals"

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Ismael Juma, Michael G. Noll, Guozhang Wang

Closes #1266 from mjsax/kafka-3599-minorCodeCleanup

Erethon committed:
MINOR: Typo fixes in ReplicaFetchMaxBytesDoc

Author: Dionysis Grigoropoulos <dgrig@erethon.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1228 from Erethon/trunk

guozhangwang committed:
KAFKA-3607: Close KStreamTestDriver upon completing; follow-up fixes to be tracked in KAFKA-3623

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Eno Thereska, Michael G. Noll, Ismael Juma

Closes #1258 from guozhangwang/K3607

Commit:
MINOR: Fix zk path in KafkaHealthCheck comment

Author: Stig <stig@linux-pr4q.suse>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1182 from srdo/comment-fix

raphw committed:
MINOR: Changed port of bootstrap server to default

Kafka is typically running on port 9092. The example named a different port what makes it difficult to run a bootstrap example without any further configuration.

Author: Rafael Winterhalter <rafael.wth@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1074 from raphw/trunk

ijuma committed:
MINOR: Upgrade to Gradle 2.13

There are a few improvements in 2.12 and 2.13. I am particularly interested in the performance improvements:
* 2.12: "This release brings support for compile only dependencies, improved build script compilation speed and even better IDE support."
* 2.13: "We've achieved performance improvements during Gradle's configuration and execution phase, where we have measured up to 25% improvements to build time in our performance tests. No changes to your build script are necessary to start taking advantage of these improvements."

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Gwen Shapira

Closes #1271 from ijuma/gradle-2.13

rajinisivaram committed:
KAFKA-3149; Extend SASL implementation to support more mechanisms

Code changes corresponding to KIP-43 to enable review of the KIP.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Jun Rao <junrao@apache.org>, Ismael Juma <ismael@juma.me.uk>

Closes #812 from rajinisivaram/KAFKA-3149

granthenke committed:
KAFKA-3306: Change metadata response to include required additional fi…

…elds

- Adds boolean type to the protocol
- Allows protocol arrays to be null (optionally)
- Adds support to ask for no topics in the metadata request
- Adds new fields to the Metadata response protocol
- Adds server code to handle new fields
   - Support no-topic metadata requests
   - Track controller id in the metadata cache
   - Check if a topic is considered internal
   - Included rack information if present
   - Include all replicas and ISRs, even if node is down
- Adds test code to test new functionality independent of the client

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Gwen Shapira, Ismael Juma, Ashish Singh

Closes #1095 from granthenke/metadata-changes

omkreddy committed:
KAFKA-3406; Update CommonClientConfigs.RETRY_BACKOFF_MS_DOC doc string

Author: Manikumar reddy O <manikumar.reddy@gmail.com>

Reviewers: Sriharsha Chintalapani <schintalapani@hortonworks.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1230 from omkreddy/KAFKA-3406

omkreddy committed:
KAFKA-3580; Improve error logging in ReplicaFetchThread

Author: Manikumar reddy O <manikumar.reddy@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1237 from omkreddy/KAFKA-3580

dguy committed:
KAFKA-3629; KStreamImpl.to(...) throws NPE when the value SerDe is null

guozhangwang

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #1272 from dguy/kstreamimpl-to-npe and squashes the following commits:

49d48fb [Damian Guy] actually commit the fix
07ce589 [Damian Guy] fix npe in KStreamImpl.to(..)
74d396d [Damian Guy] fix npe in KStreamImpl.to(..)

Commit:
KAFKA-3307; Add ApiVersions Request/Response and server side handling.

The patch does the following.
1. Adds ApiVersionsRequest/Response.
2. Adds UNSUPPORTED_VERSION error and UnsupportedVersionException.
3. Adds broker side handling of ApiVersionsRequest.

Author: Ashish Singh <asingh@cloudera.com>

Reviewers: Gwen Shapira, Ismael Juma, Magnus Edenhill

Closes #986 from SinghAsDev/KAFKA-3307

enothereska committed:
KAFKA-3612: Added structure for integration tests

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Ismael Juma, Damian Guy, Michael G. Noll, Guozhang Wang

Closes #1260 from enothereska/KAFKA-3612-integration-tests

granthenke committed:
KAFKA-3631; Fix Struct.toString for nullable arrayOf

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1279 from granthenke/struct-fix

Commit:
KAFKA-3621; Add tests for ApiVersionRequest/Response

Author: Ashish Singh <asingh@cloudera.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1275 from SinghAsDev/KAFKA-3621

Ishiihara committed:
KAFKA-3611: Remove warnings when using reflections

ewencp granders Can you take a look? Thanks!

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1259 from Ishiihara/fix-warning

rajinisivaram committed:
KAFKA-3617; Unit tests for SASL authenticator

Unit tests for SASL authenticator, tests for SASL/PLAIN and multiple mechanisms, authorization test for SASL/PLAIN

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1273 from rajinisivaram/KAFKA-3617

hachikuji committed:
KAFKA-3636; Change default max session timeout to 5 minutes

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Onur Karaman <okaraman@linkedin.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1284 from hachikuji/KAFKA-3636

Ishiihara committed:
MINOR: Avoid compiler warnings when registering metrics in KafkaServer

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1280 from Ishiihara/minor-warning-fix

Commit:
KAFKA-3382: Add system test for ReplicationVerificationTool

Author: Ashish Singh <asingh@cloudera.com>

Reviewers: Geoff Anderson <geoff@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1160 from SinghAsDev/KAFKA-3382

Commit:
KAFKA-3613: Consolidate TumblingWindows and HoppingWindows into TimeWindows

This PR includes the same code as https://github.com/apache/kafka/pull/1261 but is rebased on latest trunk.

Author: Michael G. Noll <michael@confluent.io>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #1277 from miguno/KAFKA-3613-v2

mjsax committed:
KAFKA-3598: Improve JavaDoc of public API

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Michael G. Noll, Guozhang Wang

Closes #1250 from mjsax/JavaDoc-publicAPI

rajinisivaram committed:
KAFKA-2693: Ducktape tests for SASL/PLAIN and multiple mechanisms

Run a sanity test with SASL/PLAIN and a couple of replication tests with SASL/PLAIN and multiple mechanisms.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1282 from rajinisivaram/KAFKA-2693

hachikuji committed:
KAFKA-3418: add javadoc section describing consumer failure detection

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1129 from hachikuji/KAFKA-3418

Ishiihara committed:
KAFKA-3615: Exclude test jars in kafka-run-class.sh

granders hachikuji Can you take a look when you have time? Appreciate your time to review.

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Grant Henke <granthenke@gmail.com>, Geoff Anderson <geoff@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1263 from Ishiihara/classpath-no-test-jar

apovzner committed:
KAFKA-3597; Query ConsoleConsumer and VerifiableProducer if they shutdown cleanly

Even if a test calls stop() on console_consumer or verifiable_producer, it is still possible that producer/consumer will not shutdown cleanly, and will be killed forcefully after a timeout. It will be useful for some tests to know whether a clean shutdown happened or not. This PR adds methods to console_consumer and verifiable_producer to query whether clean shutdown happened or not.

hachikuji and/or granders Please review.

Author: Anna Povzner <anna@confluent.io>

Reviewers: Jason Gustafson, Geoff Anderson, Gwen Shapira

Closes #1278 from apovzner/kafka-3597

Ishiihara committed:
KAFKA-3578: Allow cross origin HTTP requests on all HTTP methods

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1288 from Ishiihara/kip-56

rajinisivaram committed:
KAFKA-3618; Handle ApiVersionsRequest before SASL authentication

Server-side implementation and tests for handling ApiVersionsRequest before SaslHandshakeRequest.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Gwen Shapira, Ismael Juma

Closes #1286 from rajinisivaram/KAFKA-3618

mjsax committed:
KAFKA-3440: Update streams javadocs

- add class doc for KTable, KStream, JoinWindows
- add missing return tags

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Michael G. Noll <michael@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1287 from mjsax/kafka-3440-JavaDoc

granthenke committed:
KAFKA-3641; Fix RecordMetadata constructor backward compatibility

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Gwen Shapira, Ismael Juma

Closes #1292 from granthenke/recordmeta-compat

Ishiihara committed:
KAFKA-3459: Returning zero task configurations from a connector does not properly clean up existing tasks

hachikuji ewencp Can you take a look when you have time?

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1248 from Ishiihara/kafka-3459

enothereska committed:
HOTFIX: Fix equality semantics of KeyValue

Fixes wrong KeyValue equals logic when keys not equal but values equal.

Original hotfix PR at https://github.com/apache/kafka/pull/1293 (/cc enothereska)

Please review: ewencp ijuma guozhangwang

Author: Eno Thereska <eno.thereska@gmail.com>
Author: Michael G. Noll <michael@confluent.io>

Reviewers: Michael G. Noll <michael@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1294 from miguno/KeyValue-equality-hotfix

zackdever committed:
MINOR: change initial value of Min stat to Double.MAX_VALUE (not MIN)

this is consistent with the Max stat implementation.

Author: Zack Dever <zdever@pandora.com>

Reviewers: Aditya Auradkar, Gwen Shapira

Closes #1143 from zackdever/min-stat-fix

granthenke committed:
KAFKA-3644; Use Boolean protocol type for StopReplicaRequest delete_p…

…artitions

Author: Grant Henke <granthenke@gmail.com>

Reviewers: Gwen Shapira

Closes #1296 from granthenke/stop-boolean

dpkp committed:
HOTFIX: Fix main classpath libs glob for release (fixup KAFKA-3615 regression)

bin/kafka-run-class.sh does not correctly setup the CLASSPATH in release rc2.

Author: Dana Powers <dana.powers@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1302 from dpkp/KAFKA-3615-fix

Commit:
MINOR: Add version check on enable-systest-events flag

Recent patch adding enable-systest-events flag without any version check breaks all uses of versioned console consumer. E.g. upgrade tests, compatibility tests etc.

Added a check to only apply the flag if running 0.10.0 or greater.

Author: Geoff Anderson <geoff@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1298 from granders/minor-systest-fix-versioned-console-consumer

rajinisivaram committed:
KAFKA-3517; Add documentation for SASL/PLAIN

Documentation corresponding to KIP-43 - SASL/PLAIN and multiple mechanism support.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Magnus Edenhill <apache_m@edenhill.se>, Jun Rao <junrao@gmail.com>

Closes #1232 from rajinisivaram/KAFKA-3517

theduderog committed:
KAFKA-3650: Fix vagrant download URL

Author: Roger Hoover <roger.hoover@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1308 from theduderog/fix_vagrant_dl

norwood committed:
MINOR: add logfilename to error message when file missing

ewencp

Author: dan norwood <norwood@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1307 from norwood/log-filename-when-missing-file

kawamuray committed:
KAFKA-3642: Fix NPE from ProcessorStateManager when the changelog topic not exists

Issue: https://issues.apache.org/jira/browse/KAFKA-3642

Author: Yuto Kawamura <kawamuray.dadada@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1289 from kawamuray/KAFKA-3642-streams-NPE

ijuma committed:
KAFKA-3128; Add metrics for ZooKeeper events zookeeper metrics

Also:
* Remove redundant `time.milliseconds` call in `Sensor.record`
* Clean-up a number of tests and remove a manual test that is no longer required

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Liquan Pei <liquanpei@gmail.com>, Jun Rao <junrao@gmail.com>

Closes #1265 from ijuma/kafka-3128-zookeeper-metrics

Commit:
KAFKA-3448; Support zone index in IPv6 regex

When an address is written textually, the zone index is appended to the address, separated by a percent sign (%). The actual syntax of zone indices depends on the operating system.

Author: Som Sahu <sosahu@microsoft.com>
Author: Soumyajit Sahu <soumyajit-sahu@users.noreply.github.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1305 from soumyajit-sahu/fixIPV6RegexPattern_trunk

enothereska committed:
MINOR: Added more integration tests

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Ismael Juma, Michael G. Noll, Guozhang Wang

Closes #1285 from enothereska/more-integration-tests

Ishiihara committed:
MINOR: Clean up of ConsumerCoordinator and PartitionAssignor

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1306 from Ishiihara/minor-consumer-cleanup

zhuchen1018 committed:
KAFKA-3648; maxTimeToBlock in BufferPool.allocate should be enforced

 `maxTimeToBlock` needs to be updated in each loop iteration. Also record waitTime before throwing `TimeoutException`

Author: Chen Zhu <amandazhu19620701@gmail.com>

Reviewers: Dong Lin <lindong28@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1304 from zhuchen1018/KAFKA-3648

Ishiihara committed:
KAFKA-3654: Config validation should validate both common and connector specific configurations

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1313 from Ishiihara/config-short-circuit

ewencp committed:
MINOR: Handle null values in validators

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1316 from ewencp/minor-handle-null-values-validators

benstopford committed:
KAFKA-2684; Add force option to topic / config command so they can be called programatically

Tiny change to add a force option to the topic and config commands so they can be called programatically without requiring user input.

Author: Ben Stopford <benstopford@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #351 from benstopford/CPKAFKA-61B

guozhangwang committed:
MINOR: Modify checkstyle to allow import classes only used in javadoc

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Gwen Shapira, Ismael Juma

Closes #1317 from guozhangwang/KJavaDocImport

hachikuji committed:
KAFKA-3632; remove fetcher metrics on shutdown and leader migration

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1312 from hachikuji/KAFKA-3632

Commit:
KAFKA-2236; Offset request reply racing with segment rolling

Author: William Thurston <wthurston@linkedin.com>
Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Ismael Juma, Guozhang Wang

Closes #1318 from ijuma/KAFKA-2236-offset-request-reply-segment-rolling-race

Ishiihara committed:
KAFKA 3656: Remove logging outstanding messages when producer flush fails

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1319 from Ishiihara/kafka-3656

guozhangwang committed:
KAFKA-3639; Configure default serdes upon construction

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Michael G. Noll <michael@confluent.io>, Matthias J. Sax <matthias@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1311 from guozhangwang/K3639

rajinisivaram committed:
KAFKA-3652; Return error response for unsupported version of ApiVersionsRequest

Handle unsupported version of ApiVersionsRequest during SASL auth as well as normal operation by returning an error response.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Jun Rao <junrao@gmail.com>

Closes #1310 from rajinisivaram/KAFKA-3652

Ishiihara committed:
KAFKA-3527: Consumer commitAsync should not expose internal exceptions

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Grant Henke <granthenke@gmail.com>, Jason Gustafson <jason@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1300 from Ishiihara/kafka-3527

theduderog committed:
MINOR: Fixes to AWS init script for testing

A path was wrong in the script and in the documentation.

Author: Roger Hoover <roger.hoover@gmail.com>

Reviewers: Geoff Anderson <geoff@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1309 from theduderog/fix_aws_init

hachikuji committed:
KAFKA-3659: Handle coordinator disconnects more gracefully in client

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Grant Henke <granthenke@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1322 from hachikuji/KAFKA-3659

Commit:
KAFKA-3581: add timeouts to joins in background thread services

This actually removes joins altogether, as well as references to self.worker_threads, which is best left as an implementation detail in BackgroundThreadService.

This makes use of hachikuji 's recent ducktape patch, and updates ducktape dependency to 0.5.0.

Author: Geoff Anderson <geoff@confluent.io>

Reviewers: Jason Gustafson <jason@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1297 from granders/KAFKA-3581-systest-add-join-timeout

enothereska committed:
HOTFIX: Reverted timeouts to larger values

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Ismael Juma, Guozhang Wang

Closes #1324 from enothereska/hotfix-timeouts

onurkaraman committed:
KAFKA-3661; fix NPE in o.a.k.c.c.RoundRobinAssignor when topic metadata not found

AbstractPartitionAssignor.assign has an ambiguous line in its documentation:
> param partitionsPerTopic The number of partitions for each subscribed topic (may be empty for some topics)

Does empty mean the topic has an entry with value zero, or that the entry is excluded from the map altogether? The current implementation in AbstractPartitionAssignor excludes the entry from partitionsPerTopic if the topic isn't in the metadata.

RoundRobinAssignorTest.testOneConsumerNonexistentTopic interprets emptiness as providing the topic with a zero value.
RangeAssignor interprets emptiness as excluding the entry from the map.
RangeAssignorTest.testOneConsumerNonexistentTopic interprets emptiness as providing the topic with a zero value.

This implementation chooses to solve the NPE by deciding to exclude topics from partitionsPerTopic when the topic is not in the metadata.

Author: Onur Karaman <okaraman@linkedin.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1326 from onurkaraman/KAFKA-3661

MayureshGharat committed:
KAFKA-3651; Remove the condition variable waiting on memory availability in Bufferpool when a TimeoutException is thrown

Whenever the BufferPool throws a "Failed to allocate memory within the configured max blocking time" exception, it should also remove the condition object from the waiters deque

Author: MayureshGharat <gharatmayuresh15@gmail.com>

Reviewers: Chen Zhu <amandazhu19620701@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1314 from MayureshGharat/kafka-3651

guozhangwang committed:
HOTFIX: follow-up on KAFKA-725 to remove the check and return empty response instead of throw exceptions

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Stig Døssing, Ismael Juma, Jun Rao

Closes #1327 from guozhangwang/K725r

hachikuji committed:
KAFKA-3627: consumer fails to execute delayed tasks in poll when records are available

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Liquan Pei <liquanpei@gmail.com>, Jiangjie Qin <becket.qin@gmail.com>, Guozhang Wang <wangguoz@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1295 from hachikuji/KAFKA-3627

Commit:
KAFKA-3666; Update links for new consumer API

Pull request to update the consumer API links in the docs.

Author: Dustin Cote <dustin@dustins-mbp.attlocal.net>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1331 from cotedm/KAFKA-3666

kawamuray committed:
KAFKA-3616: Make kafka producers/consumers injectable for KafkaStreams

Ticket: https://issues.apache.org/jira/browse/KAFKA-3616

Author: Yuto Kawamura <kawamuray.dadada@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1264 from kawamuray/kafka-3616-inject-clients

Commit:
KAFKA-3592: System test - configurable paths

This patch adds logic for the following:
- remove hard-coded paths to various scripts and jars in kafkatest service classes
- provide a mechanism for overriding path resolution logic with a "pluggable" path resolver class

Author: Geoff Anderson <geoff@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1245 from granders/configurable-install-path

coughman committed:
Minor: added description for lag in consumer group command

Author: Kaufman Ng <kaufman@confluent.io>

Reviewers: Gwen Shapira

Closes #1320 from coughman/trunk

markgrover committed:
KAFKA-3669; Add secondary constructor for KafkaConfig with a default value for doLog

…value for doLog

Author: Mark Grover <mark@apache.org>

Reviewers: Grant Henke <granthenke@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1334 from markgrover/kafka-3669-trunk

zhuchen1018 committed:
KAFKA-3655; awaitFlushCompletion() in RecordAccumulator should always decrement flushesInProgress count

Author: Chen Zhu <amandazhu19620701@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1315 from zhuchen1018/KAFKA-3655

JonathanBond committed:
KAFKA-3112; Warn instead of error on unresolvable bootstrap server

so that unresolvable DNS names are ignored and only throw an error if no other bootstrap servers are resolvable.

Author: Jonathan Bond <jbond@netflix.com>

Reviewers: Ewen Cheslack-Postava <me@ewencp.org>, Grant Henke <granthenke@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #792 from bondj/KAFKA-3112

dpkp committed:
KAFKA-3160; Fix LZ4 Framing

This contribution is my original work and I license the work under Apache 2.0.

Author: Dana Powers <dana.powers@gmail.com>
Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1212 from dpkp/KAFKA-3160

ijuma committed:
KAFKA-3670; ControlledShutdownLeaderSelector should pick the preferred replica as the new leader, if possible

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #1338 from ijuma/kafka-3670-controlled-shutdown-leader-selector-preferred-replica

vahidhashemian committed:
KAFKA-3579; Reference both old and new consumer properties in `TopicCommand`

Add references to the new consumer property 'max.partition.fetch.bytes' along with the old consumer property 'fetch.message.max.bytes' in the corresponding warning messages of TopicCommand.
Also, create and leverage a static variable for the default value of the new consumer property.
Also, use 'DEFAULT_...' for default propoerty constant names in the code instead of '..._DEFAULT'.

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Manikumar reddy O <manikumar.reddy@gmail.com>, Ashish Singh <asingh@cloudera.com>, Grant Henke <granthenke@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1239 from vahidhashemian/KAFKA-3579

Ishiihara committed:
KAFKA 3671: Move topics to SinkConnectorConfig

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Dan Norwood <norwood@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1335 from Ishiihara/sink-connector-config

Ishiihara committed:
KAFKA-3673: Connect tests don't handle concurrent config changes

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1340 from Ishiihara/connect-test-failure

Ishiihara committed:
MINOR: Move connect.start() to try catch block

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1347 from Ishiihara/connect-standalone

hachikuji committed:
KAFKA-3674: Ensure connector target state changes propagated to worker

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1341 from hachikuji/KAFKA-3674

ijuma committed:
KAFKA-3675; Add lz4 to parametrized `test_upgrade` system test

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1343 from ijuma/kafka-3675-lz4-test-upgrade

rajinisivaram committed:
KAFKA-3662; Fix timing issue in SocketServerTest.tooBigRequestIsRejected

Test sends large request using multiple writes of length followed by request body. The first write should succeed, but since the server closes the connection on processing the length that is too big, subsequent writes may fail. Modified test to handle this exception.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1349 from rajinisivaram/KAFKA-3662

coughman committed:
KAFKA-3681; Connect doc formatting

Author: Kaufman Ng <kaufman@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1351 from coughman/KAFKA-3681-connect-doc-formatting

guozhangwang committed:
KAFKA-3658: Validate retention period be longer than window size

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Henry Cai, Ismael Juma

Closes #1337 from guozhangwang/K3658

mimaison committed:
KAFKA-3587; LogCleaner fails due to incorrect offset map computation

Removed the over pessimistic require and instead attempt to fill the dedup buffer. Use the (only) map until full;
this may allow to process all dirty segment (optimism) or may happen in the middle of a dirt segment.
In either case, do compaction using the map loaded that way.

This patch was developed with edoardocomar

Author: Mickael Maison <mickael.maison@gmail.com>

Reviewers: Jun Rao <junrao@gmail.com>, Guozhang Wang <wangguoz@gmail.com>

Closes #1332 from mimaison/KAFKA-3587

dpkp committed:
HOTFIX: KAFKA-3160 follow-up, catch decompression errors in constructor

After testing KAFKA-3160 a bit more, I found that the error code was not being set properly in ProduceResponse. This happened because the validation error is raised in the CompressionFactory constructor, which was not wrapped in a try / catch.

ijuma junrao

(This contribution is my original work and I license the work under Apache 2.0.)

Author: Dana Powers <dana.powers@gmail.com>
Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jun Rao <junrao@gmail.com>, Gwen Shapira <cshapi@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1344 from dpkp/decompress_error_code

hachikuji committed:
MINOR: Add toString implementations to Subscription and Assignment

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1354 from hachikuji/minor-add-missing-assignor-tostrings

rajinisivaram committed:
MINOR: Remove copyDependantTestLibs from jar dependencies

_copyDependantTestLibs_ was added temporarily as a dependency of _jar_ task to enable SASL system tests to be run with MiniKdc without changing the automated system test runs which run _gradlew clean jar_. Since the build target _systemTestLibs_ is already in Kafka build.gradle, the Confluent automated test runs can now run _gradlew clean systemTestLibs_ instead. This PR provides the final change to remove _copyDependantTestLibs_ from the _jar_ task. This should be committed only after the Confluent automated sytem test build script is updated, to avoid breaking any builds.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #430 from rajinisivaram/minor-systemtestlibs

ewencp committed:
KAFKA-3421; Follow up to fix name of SourceTask method and add documentation of connector status REST API

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1355 from ewencp/kafka-3421-follow-up

rajinisivaram committed:
KAFKA-3634; Upgrade tests for SASL authentication

Add a test for changing SASL mechanism using rolling upgrade and a test for rolling upgrade from 0.9.0.x to 0.10.0 with SASL/GSSAPI.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ben Stopford <benstopford@gmail.com>, Geoff Anderson <geoff@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1290 from rajinisivaram/KAFKA-3634

Commit:
MINOR: Update protocol doc link in Introduction.

Author: Ashish Singh <asingh@cloudera.com>

Reviewers: Gwen Shapira

Closes #1211 from SinghAsDev/MinorFixDocLink

hachikuji committed:
KAFKA-3676: system tests for connector pause/resume

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1345 from hachikuji/KAFKA-3676

vahidhashemian committed:
KAFKA-3608; Fix ZooKeeper structures and output format in documentation

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Gwen Shapira

Closes #1257 from vahidhashemian/KAFKA-3608

Ishiihara committed:
KAFKA-3684: SinkConnectorConfig does not return topics in config validation.

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1356 from Ishiihara/bug-fix-validate

guozhangwang committed:
MINOR: Add Kafka Streams API / upgrade notes

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Michael G. Noll <michael@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1321 from guozhangwang/KStreamsJavaDoc

ijuma committed:
MINOR: Double timeout passed to `producer.close` in `sendAndVerifyTimestamp`

We have had transient failures in this method when Jenkins is
overloaded.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1359 from ijuma/increase-producer-close-timeout-in-test

omkreddy committed:
KAFKA-3525; getSequenceId should return 1 for first data node creation

ZkUtils.getSequenceId() method is used to generate broker id sequence numbers. During startup, each broker updates the data at /brokers/seqid zk path and returns stat.getVersion as next sequence id.

stat.getVersion returns "1" for first data update. So ZkUtils.getSequenceId() should return "1" on first data update.

Author: Manikumar reddy O <manikumar.reddy@gmail.com>

Reviewers: Flavio Junqueira <fpj@apache.org>, Ismael Juma <ismael@juma.me.uk>

Closes #1224 from omkreddy/KAFKA-3525

rajinisivaram committed:
MINOR: Ensure that selection key is cancelled on close

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1368 from rajinisivaram/minor-channelclose

benstopford committed:
MINOR: Documentation for Rack Awareness

Author: Ben Stopford <benstopford@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1369 from benstopford/rack-awareness-docs

Ishiihara committed:
KAFKA-3690: Avoid to pass null to UnmodifiableMap

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1360 from Ishiihara/avoid-to-pass-null

rajinisivaram committed:
MINOR: Improve handling of channel close exception

Propagate IOException in SslTransportLayer channel.close to be consistent with PlaintextTransportLayer,  close authenticator on channel close even if transport layer close fails.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1370 from rajinisivaram/minor-channelclose2

hachikuji committed:
KAFKA-3694; Ensure broker Zk deregistration prior to restart in ReplicationTest

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Geoff Anderson <geoff@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1365 from hachikuji/KAFKA-3694

guozhangwang committed:
KAFKA-3704: Remove hard-coded block size in KafkaProducer

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Ismael Juma

Closes #1371 from guozhangwang/K3565-remove-compression-blocksize

hachikuji committed:
KAFKA-3583: Add documentation for Connect status control APIs

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Roger Hoover <roger.hoover@gmail.com>, Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1358 from hachikuji/KAFKA-3583

ijuma committed:
MINOR: Fix order of compression algorithms in upgrade note

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Jun Rao <junrao@gmail.com>

Closes #1373 from ijuma/fix-producer-buffer-size-upgrade-note

guozhangwang committed:
MINOR: Change type of StreamsConfig.BOOTSTRAP_SERVERS_CONFIG to List

This is an improved version of https://github.com/apache/kafka/pull/1374, where we include a unit test.

/cc ijuma and guozhangwang

Author: Guozhang Wang <wangguoz@gmail.com>
Author: Michael G. Noll <michael@confluent.io>

Reviewers: Michael G. Noll <michael@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1377 from miguno/streamsconfig-multiple-bootstrap-servers

ijuma committed:
MINOR: Fix checkstyle failure in `StreamsConfigTest`

I removed the hamcrest matcher to unbreak the build, but we probably want to tweak the `import-control.xml` as it currently only allows it for `<subpackage name="integration">`, which is weird.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1380 from ijuma/fix-streams-config-test-checkstyle

ijuma committed:
MINOR: Remove <release> tag from doap file

The information from this file is displayed in:

https://projects.apache.org/project.html?kafka

I removed the tag instead of fixing it because it's one more thing to maintain and neither Hadoop nor ZooKeeper have it (httpd does though).

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1378 from ijuma/remove-release-from-doap

Ishiihara committed:
KAFKA-3421: Connect developer guide update and several fixes

This is a follow up of KAKFA-3421 to update the connect developer guide to include the configuration validation. Also includes a couple of minor fixes.

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1366 from Ishiihara/connect-dev-doc

Ishiihara committed:
KAFKA-3520: Add system tests for REST APIs of list connector plugins and config validation

ewen granders Ready for review.

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1195 from Ishiihara/system-test

jklukas committed:
MINOR: Fix bugs in KafkaStreams.close()

Initially proposed by ijuma in https://github.com/apache/kafka/pull/1362#issuecomment-218293662

mjsax commented:

> StreamThread.close() should be extended to call metrics.close() (the class need a private member to reference the Metrics object, too)

The `Metrics` instance is created in the `KafkaStreams` constructor and shared between all threads, so closing it within the threads doesn't seem like the right approach. This PR calls `Metrics.close()` in `KafkaStreams.close()` instead.

cc guozhangwang

Author: Jeff Klukas <jeff@klukas.net>

Reviewers: Ismael Juma, Guozhang Wang

Closes #1379 from jklukas/close-streams-metrics

Ishiihara committed:
KAFKA-3692; Add quotes to variables in kafka-run-class.sh

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Geoff Anderson <geoff@confluent.io>, Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1364 from Ishiihara/add-quote-classpath

omkreddy committed:
KAFKA-3584; Fix synchronization issue between deleteOldSegments() and delete() methods

This PR is to fix synchronization issue between deleteOldSegments() and delete() method calls. log.deleteOldSegments() call throws NullPointerException after log.delete() method call.

cc ijuma junrao

Author: Manikumar reddy O <manikumar.reddy@gmail.com>

Reviewers: Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1367 from omkreddy/KAFKA-3584

jklukas committed:
MINOR: Allow Serdes subclasses to access WrapperSerde

This PR loosens access restrictions on `WrapperSerde` to `protected` so that users can define a `Serdes` subclass that provides additional custom serde members following the same pattern as the parent class.

This is my own work and is compatible with Kafka's license.

Author: Jeff Klukas <jeff@klukas.net>

Reviewers: Guozhang Wang

Closes #1382 from jklukas/expose-wrapperserde

ijuma committed:
KAFKA-3713; Close `compressor` to fix memory leak

This fixes test_producer_throughput with compression_type=snappy.

Also: added heap dump on out of memory error to `producer_performance.py` and corrected the upgrade note related to the change in buffer size for compression streams.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Gwen Shapira

Closes #1385 from ijuma/kafka-3713-test_producer_throughput-snappy-fail and squashes the following commits:

54c7962 [Ismael Juma] Correct upgrade note about buffer size for compression stream
515040b [Ismael Juma] Call `compressor.close()` to fix memory leak
5311e5b [Ismael Juma] Dump heap on out of memory error when running `producer_performance.py`

becketqin committed:
KAFKA-3565; Add documentation to warn user about the potential messag…

…e throughput drop due to the addition of timestamp field.

Author: Jiangjie Qin <becket.qin@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Jun Rao <junrao@gmail.com>

Closes #1372 from becketqin/KAFKA-3565

tom-pang committed:
MINOR: document increased network bandwidth of 0.10 under replication

If you're pushing close to the network capacity, 0.10's additional 8 bytes per message can lead to overload of your network. We (Heroku Kafka) ran into this issue whilst benchmarking 0.10 RC and the ijuma suggested it belonged in the update note.

Comments/suggestions welcome.

Author: Tom Crayford <tcrayford@googlemail.com>

Reviewers: Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1389 from tcrayford/upgrade_note_about_increased_network_bandwidth

rajinisivaram committed:
KAFKA-3258; Delete broker topic metrics of deleted topics

Delete per-topic metrics when there are no replicas of any partitions of the topic on a broker.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Joel Koshy <jjkoshy.w@gmail.com>, Manikumar reddy O <manikumar.reddy@gmail.com>, Ashish Singh <asingh@cloudera.com, Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #944 from rajinisivaram/KAFKA-3258

becketqin committed:
KAFKA-3698; Update the message format section.

Author: Jiangjie Qin <becket.qin@gmail.com>

Reviewers: Jun Rao <junrao@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1375 from becketqin/KAFKA-3698

ijuma committed:
KAFKA-3704; Revert "Remove hard-coded block size in KafkaProducer"

This is not an exact revert as the code changed a bit since the
original commit. We also include a note in `upgrade.html`.

The original commit is 1182d61deb23b5cd86cbe462471f7df583a796e1.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Gwen Shapira, Guozhang Wang

Closes #1391 from ijuma/kafka-3704-revert and squashes the following commits:

7891b67 [Ismael Juma] Tweak upgrade note based on Gwen's feedback
1673cd0 [Ismael Juma] Revert "KAFKA-3704: Remove hard-coded block size in KafkaProducer"

guozhangwang committed:
MINOR: Add INFO logging if ZK config is not specified

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Ismael Juma

Closes #1392 from guozhangwang/Kminor-warn-no-zk-config

guozhangwang committed:
KAFKA-3716; Validate all timestamps are not negative

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Eno Thereska, Ismael Juma

Closes #1393 from guozhangwang/K3716-check-non-negative-timestamps

MayureshGharat committed:
KAFKA-3393; Updated the docs to reflect the deprecation of block.on.buffer.full and usage of max.block.ms

Author: MayureshGharat <gharatmayuresh15@gmail.com>

Reviewers: Grant Henke <granthenke@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1060 from MayureshGharat/KAFKA-3393

ijuma committed:
MINOR: Exclude jline, netty and findbugs annotations

These dependencies are unnecessary and they are acquired
transitively via zkclient (jline, netty) and reflections (annotations).

Ewen did the hard work in figuring out why we have unexpected
additional dependencies since 0.9.x.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Ewen Cheslack-Postava, Guozhang Wang, Gwen Shapira

Closes #1396 from ijuma/exclude-jline-netty-deps-in-streams and squashes the following commits:

3aa366f [Ismael Juma] Exclude findbugs annotations due to LGPL license
2d3d714 [Ismael Juma] Use local exclusion for `jline` and `netty`
482b6c0 [Ismael Juma] Exclude `jline` and `netty` dependencies in the `streams` project

becketqin committed:
KAFKA-3721; Put UpdateMetadataRequest V2 in 0.10.0-IV1

Author: Jiangjie Qin <becket.qin@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Gwen Shapira <cshapi@gmail.com>, Jun Rao <junrao@gmail.com>

Closes #1400 from becketqin/KAFKA-3721

ijuma committed:
MINOR: Use `Record` instead of `ByteBufferMessageSet` in `ProduceRequestTest`

We want to phase out `ByteBufferMessageSet` eventually, so new code should favour `Record` where possible.

Also use a fixed timestamp in `testCorruptLz4ProduceRequest` to ensure that
the checksum is always the same.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jason Gustafson, Guozhang Wang

Closes #1357 from ijuma/produce-request-test-improvement

Commit:
MINOR: Bump system test ducktape dependency to 0.5.1

Author: Geoff Anderson <geoff@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1397 from granders/minor-increment-ducktape

ijuma committed:
KAFKA-3717; Support building aggregate javadoc for all project modules

The task is called `aggregatedJavadoc` and the generated html will be under `<project.dir>/build/docs/javadoc/`.

I also disabled javadoc for `tools` and `log4j-appender` as they are not public API.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1398 from ijuma/kafka-3717-aggregate-javadoc

ijuma committed:
MINOR: Move `Incorporating Security Features in a Running Cluster` to its own section under `Security`

It was previously in the SASL section (probably by mistake).

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Sriharsha Chintalapani <schintalapani@hortonworks.com>

Closes #1405 from ijuma/fix-security-upgrade-location-in-docs

shikhar committed:
KAFKA-2935: Remove vestigial WorkerConfig.CLUSTER_CONFIG

Author: shikhar <shikhar@schmizz.net>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1404 from shikhar/kafka-2935

guozhangwang committed:
KAFKA-3735: Dispose all RocksObejcts upon completeness

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Roger Hoover, Eno Thereska, Ismael Juma

Closes #1411 from guozhangwang/K3735-dispose-rocksobject

Commit:
MINOR: Specify keyalg RSA for SSL key generation

Author: Sriharsha Chintalapani <harsha@hortonworks.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1416 from harshach/ssl-doc-fix

ijuma committed:
KAFKA-3747; Close `RecordBatch.records` when append to batch fails

With this change, `test_producer_throughput` with message_size=10000, compression_type=snappy and a snappy buffer size of 32k can be executed in a heap of 192m in a local environment (768m is needed without this change).

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1418 from ijuma/kafka-3747-close-record-batch-when-append-fails

ijuma committed:
MINOR: Fix documentation table of contents and `BLOCK_ON_BUFFER_FULL_DOC`

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Gwen Shapira

Closes #1423 from ijuma/minor-doc-fixes

Commit:
Minor: Fix ps command example in docs

Process grep command has been updated. Previous "ps  | grep server-1.properties"  command is showing nothing.

Author: Satendra Kumar <satendra@knoldus.com>

Reviewers: Gwen Shapira

Closes #1386 from satendrakumar06/patch-1

Commit:
KAFKA-3683; Add file descriptor recommendation to ops guide

Adding sizing recommendations for file descriptors to the ops guide.

Author: Dustin Cote <dustin@confluent.io>
Author: Dustin Cote <dustin@dustins-mbp.attlocal.net>

Reviewers: Gwen Shapira

Closes #1353 from cotedm/KAFKA-3683 and squashes the following commits:

8120318 [Dustin Cote] Adding file descriptor sizing recommendations
0908aa9 [Dustin Cote] Merge https://github.com/apache/kafka into trunk
32315e4 [Dustin Cote] Merge branch 'trunk' of https://github.com/cotedm/kafka into trunk
13309ed [Dustin Cote] Update links for new consumer API
4dcffc1 [Dustin Cote] Update links for new consumer API

Ishiihara committed:
MINOR: Add virtual env to Kafka system test README.md

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Gwen Shapira

Closes #1346 from Ishiihara/add-venv

enothereska committed:
MINOR: Removed 1/2 of the hardcoded sleeps in Streams

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1422 from enothereska/minor-integration-timeout2

mimaison committed:
KAFKA-3732: Add an auto accept option to kafka-acls.sh

Added a new argument to AclCommand: --yes. When set, automatically answer yes to prompts

Author: Mickael Maison <mickael.maison@gmail.com>

Reviewers: Gwen Shapira

Closes #1406 from mimaison/KAFKA-3732

onurkaraman committed:
KAFKA-3718; propagate all KafkaConfig __consumer_offsets configs to OffsetConfig instantiation

Kafka has two configurable compression codecs: the one used by the client (source codec) and the one finally used when storing into the log (target codec). The target codec defaults to KafkaConfig.compressionType and can be dynamically configured through zookeeper.

The GroupCoordinator appends group membership information into the __consumer_offsets topic by:
1. making a message with group membership information
2. making a MessageSet with the single message compressed with the source codec
3. doing a log.append on the MessageSet

Without this patch, KafkaConfig.offsetsTopicCompressionCodec doesn't get propagated to OffsetConfig instantiation, so GroupMetadataManager uses a source codec of NoCompressionCodec when making the MessageSet. Let's say we have enough group information such that the message formed exceeds KafkaConfig.messageMaxBytes before compression but would fall below the threshold after compression using our source codec. Even if we had dynamically configured __consumer_offsets with our favorite compression codec, the log.append will throw RecordTooLargeException during analyzeAndValidateMessageSet since the message was unexpectedly uncompressed instead of having been compressed with the source codec defined by KafkaConfig.offsetsTopicCompressionCodec.

Author: Onur Karaman <okaraman@linkedin.com>

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1394 from onurkaraman/KAFKA-3718

theduderog committed:
Setting broker state as running after publishing to ZK

junrao

Currently, the broker state is set to running before it registers itself in ZooKeeper.  This is too early in the broker lifecycle.  If clients use the broker state as an indicator that the broker is ready to accept requests, they will get errors.  This change is to delay setting the broker state to running until it's registered in ZK.

Author: Roger Hoover <roger.hoover@gmail.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #1426 from theduderog/broker-running-after-zk

ijuma committed:
MINOR: Use `--force` instead of `--yes` in `AclCommand`

To be consistent with `ConfigCommand` and `TopicCommand`.

No release includes this option yet, so we can simply change it.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Mickael Maison, Grant Henke

Closes #1430 from ijuma/use-force-instead-of-yes-in-acl-command and squashes the following commits:

bdf3a57 [Ismael Juma] Update `AclCommandTest`
78b8467 [Ismael Juma] Change variable name to `forceOpt`
0bb27af [Ismael Juma] Use `--force` instead of `--yes` in `AclCommand`

gyk committed:
MINOR: Fix wrong comments

Author: Yukun Guo <gyk.net@gmail.com>

Reviewers: Gwen Shapira

Closes #1198 from gyk/fix-comment

christian-posta committed:
KAFKA-3723: Cannot change size of schema cache for JSON converter

Author: Christian Posta <christian.posta@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1401 from christian-posta/ceposta-connect-class-cast-error

Commit:
KAFKA-3710: MemoryOffsetBackingStore shutdown

ExecutorService needs to be shutdown on close, lest a zombie thread
prevent clean shutdown.

ewencp

Author: Peter Davis <peter.davis@expeditors.com>

Reviewers: Liquan Pei <liquanpei@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1383 from davispw/KAFKA-3710

leisore committed:
MINOR: Delete unused code in FileStreamSourceTask

Author: leisore <leisore@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1433 from leisore/master

manuzhang committed:
KAFKA-3749; fix "BOOSTRAP_SERVERS_DOC" typo

Author: manuzhang <owenzhang1990@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1420 from manuzhang/KAFKA-3749

mimaison committed:
MINOR: Fix tracing in KafkaApis.handle()

requestObj() returns null for the o.a.k.c.requests objects so use header() for these.

Once all the requests will have been replaced by o.a.k.c.requests objects, we should be able to clean that up, but in the meantime it's useful to trace both.

Author: Mickael Maison <mickael.maison@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1435 from mimaison/kafkaapis_trace

vahidhashemian committed:
MINOR: Fix a couple of scaladoc typos

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1440 from vahidhashemian/typo06/fix_typos_in_code_comments

edoardocomar committed:
KAFKA-3682; ArrayIndexOutOfBoundsException thrown by

SkimpyOffsetMap.get() when full

Limited number of attempts to number of map slots after the internal
positionOf() goes into linear search mode.
Added unit test

Co-developed with mimaison

Author: edoardo <ecomar@uk.ibm.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #1352 from edoardocomar/KAFKA-3682

enothereska committed:
KAFKA-3678: Removed sleep from streams integration tests

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1439 from enothereska/KAFKA-3678-timeouts1

sasakitoa committed:
KAFKA-3767; Add missing license to connect-test.properties

This address to https://issues.apache.org/jira/browse/KAFKA-3767.

Author: Sasaki Toru <sasakitoa@nttdata.co.jp>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1443 from sasakitoa/test_failure_no_license

imandhan committed:
KAFKA-3158; ConsumerGroupCommand should tell whether group is actually dead

This patch fix differentiates between when a consumer group is rebalancing or dead and reports the appropriate error message.

Author: Ishita Mandhan <imandha@us.ibm.com>

Reviewers: Vahid Hashemian <vahidhashemian@us.ibm.com>, Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1429 from imandhan/KAFKA-3158

rekhajoshm committed:
KAFKA-3765; Kafka Code style corrections

Removed explicit returns, not needed parentheses, corrected variables, removed unused imports
Using isEmpty/nonEmpty  instead of size check, using head, flatmap instead of map-flatten

Author: Joshi <rekhajoshm@gmail.com>
Author: Rekha Joshi <rekhajoshm@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1442 from rekhajoshm/KAFKA-3765

Ishiihara committed:
MINOR: Remove synchronized as the tasks are executed sequentially

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1441 from Ishiihara/remove-synchronized

harshach commented:
@gfodor can you add JIRA title to your PR,  like "KAFKA-3770: KStream job should be able to specify linger.ms"

ijuma committed:
MINOR: Avoid trace logging computation in `checkEnoughReplicasReachOffset`

`numAcks` is only used in the `trace` logging statement so it should be a `def` instead of a `val`. Also took the chance to improve the code and documentation a little.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1449 from ijuma/minor-avoid-trace-logging-computation-in-partition

guozhangwang commented:
Thanks @gfodor , LGTM overall. Could you add a test case in `StreamsConfigTest` for validating that underlying clients' specific configs like `ProducerConfig.LINGER_MS_CONFIG` and `ConsumerConfig.AUTO_OFFSET_RESET_CONFIG` are successfully passed via `StreamsConfig`?

coughman committed:
MINOR: added spacing in streams doc in section 9.2

Author: Kaufman Ng <kaufman@confluent.io>

Reviewers: Guozhang Wang

Closes #1454 from coughman/trunk

reftel committed:
MINOR; Close parenthesis in PartitionInfo.toString

Trivial patch
is trivial.

Author: Magnus Reftel <magnus.reftel@skatteetaten.no>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1458 from magnusr/feature/paren

fpj committed:
KAFKA-3660: Log exception message in ControllerBrokerRequestBatch

Author: Flavio Junqueira <fpj@apache.org>

Reviewers: Grant Henke <granthenke@gmail.com>, Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1325 from fpj/KAFKA-3660

izeye committed:
MINOR: Add a newline to a Javadoc sample in KafkaConsumer

This PR simply adds a newline to a Javadoc sample in `KafkaConsumer` to flush the outputs.

Author: Johnny Lim <izeye@naver.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1390 from izeye/patch-1

vahidhashemian committed:
KAFKA-3774: Make 'time' an optional argument of GetOffsetShell

Since the 'time' argument has a default value of -1, it makes sense to make it an optional argument.

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1457 from vahidhashemian/KAFKA-3774

mjsax committed:
HOTFIX: updated JavaDoc example for 0.9 tech-prev to 0.10

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Sriharsha Chintalapani, Ismael Juma, Guozhang Wang

Closes #1450 from mjsax/hotfix

ijuma committed:
MINOR: Fix setting of ACLs and ZK shutdown in test harnesses

I found both issues while investigating the issue described in PR #1425.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Sriharsha Chintalapani <schintalapani@hortonworks.com>, Jun Rao <junrao@gmail.com>

Closes #1455 from ijuma/fix-integration-test-harness-and-zk-test-harness

gfodor committed:
Remove unnecessary import

gfodor commented:
Yeah, I screwed this up when re-basing, will open a new one.
